File: app/src/main/java/se/jakob/knarkklocka/data/Alarm.java
package se.jakob.knarkklocka.data;

import android.arch.persistence.room.ColumnInfo;
import android.arch.persistence.room.Entity;
import android.arch.persistence.room.Ignore;
import android.arch.persistence.room.PrimaryKey;
import android.support.annotation.NonNull;

import java.util.Date;

@Entity(tableName = "alarm_table")
public class Alarm {

    public static final int STATE_WAITING = 0;
    public static final int STATE_DEAD = 1;
    public static final int STATE_ACTIVE = 2;
    public static final int STATE_SNOOZING = 3;

    @PrimaryKey(autoGenerate = true)
    private int id;
    @ColumnInfo(name = "alarm_state")
    private int state;
    @NonNull
    @ColumnInfo(name = "start_time")
    private Date startTime;
    @NonNull
    @ColumnInfo(name = "end_time")
    private Date endTime;
    @ColumnInfo(name = "number_of_snoozes")
    private int snoozes;

    @Ignore
    public Alarm(int state, @NonNull Date startTime, @NonNull Date endTime) {
        this.state = state;
        this.startTime = startTime;
        this.endTime = endTime;
        this.snoozes = 0;
    }

    public Alarm(int id, int state, @NonNull Date startTime, @NonNull Date endTime) {
        this.id = id;
        this.state = state;
        this.startTime = startTime;
        this.endTime = endTime;
        this.snoozes = 0;
    }

    public int getState() {return state;}

    public void setState(int state) {this.state = state;}

    public int getId() {return id;}

    public void setId(int id) {this.id = id;}

    @NonNull
    public Date getStartTime() {return startTime;}

    public void setStartTime(@NonNull Date startTime) {this.startTime = startTime;}

    @NonNull
    public Date getEndTime() {return endTime;}

    public void setEndTime(@NonNull Date endTime) {this.endTime = endTime;}

    public int getSnoozes() {return snoozes;}

    public void setSnoozes(int snoozes) {this.snoozes = snoozes;}

    public void incrementSnoozes() {
        this.snoozes += 1;
    }

    public String getStateString(int state) {
        switch (state) {
            case Alarm.STATE_ACTIVE:
                return "Active";
            case Alarm.STATE_WAITING:
                return "Waiting";
            case Alarm.STATE_DEAD:
                return "Dead";
            case Alarm.STATE_SNOOZING:
                return "Snoozed";
            default:
                return "Invalid State";
        }
    }
}

File: app/src/main/java/se/jakob/knarkklocka/data/Alarm.kt
package se.jakob.knarkklocka.data


import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.PrimaryKey
import se.jakob.knarkklocka.BuildConfig
import java.lang.Exception
import java.util.*

@Entity(tableName = "alarm_table")
data class Alarm constructor(
        @PrimaryKey(autoGenerate = true) var id: Long,
        @ColumnInfo(name = "alarm_state") var state: AlarmState = AlarmState.STATE_WAITING,
        @ColumnInfo(name = "start_time") var startTime: Date,
        @ColumnInfo(name = "end_time") var endTime: Date,
        @ColumnInfo(name = "number_of_snoozes") var snoozes: Int = 0
) {

    val waiting : Boolean
    get () {
        return state == AlarmState.STATE_WAITING
    }
    val active : Boolean
        get () {
            return state == AlarmState.STATE_ACTIVE
        }
    val snoozing : Boolean
        get () {
            return state == AlarmState.STATE_SNOOZING
        }
    val dead : Boolean
        get () {
            return state == AlarmState.STATE_DEAD
        }
    val missed : Boolean
        get() {
            return state == AlarmState.STATE_MISSED
        }

    @Ignore
    constructor(state: AlarmState, startTime: Date, endTime: Date) : this(id = 0, state = state, startTime = startTime, endTime = endTime)

    @Ignore
    constructor(startTime: Date, endTime: Date) : this(id = 0, startTime=startTime, endTime=endTime)

    val stateToString: String
        get() {
            return when (this.state) {
                AlarmState.STATE_ACTIVE -> "Active"
                AlarmState.STATE_WAITING -> "Waiting"
                AlarmState.STATE_DEAD -> "Dead"
                AlarmState.STATE_SNOOZING -> "Snoozed"
                AlarmState.STATE_MISSED -> "Missed"
            }
        }

    fun activate() {
        if (dead or active) {
            if (BuildConfig.DEBUG) {
                throw InvalidStateChangeException("Activating a dead or active Alarm is not allowed.")
            }
        } else {
            this.state = AlarmState.STATE_ACTIVE
        }
    }

    fun snooze(newEndTime: Date) {
        if (!(active or missed)) {
            if (BuildConfig.DEBUG) {
                throw InvalidStateChangeException("Only an active or missed Alarm can be snoozed.")
            }
        } else {
            state = AlarmState.STATE_SNOOZING
            incrementSnoozeCount()
            endTime = newEndTime
        }
    }

    fun kill() {
        this.endTime = Calendar.getInstance().time
        this.state = AlarmState.STATE_DEAD
    }

    fun miss() {
        if (!(active or snoozing)) {
            if (BuildConfig.DEBUG) {
                throw InvalidStateChangeException("Only an active Alarm can be missed.")
            }
        } else {
            this.state = AlarmState.STATE_MISSED
        }
    }

    private fun incrementSnoozeCount() {
        this.snoozes += 1
    }

    override fun toString(): String {
        return String.format(Locale.getDefault(), "Alarm \n id: %d \n startTime: %s \n endTime: %s \n state: %s",
                id, startTime.toString(), endTime.toString(), stateToString)
    }

    class InvalidStateChangeException(message:String) : Exception(message)
}

